src/components/tl/ucp/allreduce/allreduce_sliding_window.c

```diff
@@ -11,80 +11,507 @@
-    size_t __attribute__((unused)) put_window_size) //NOLINT
+static inline void ucc_tl_ucp_allreduce_sliding_window_reset_pipeline(
+    ucc_tl_ucp_allreduce_sw_pipeline_t *pipe, ucc_rank_t rank,
+    size_t put_window_size)
 {
+    int i;
+
+    pipe->avail_buffs   = pipe->num_buffers;
+    pipe->src_rank      = pipe->dst_rank       = rank;
+    pipe->get_idx       = pipe->red_idx        = 0;
+    pipe->done_get      = pipe->done_red       = 0;
+    pipe->done_put      = pipe->posted_put     = 0;
+    pipe->count_issued  = pipe->count_received = 0;
+    pipe->count_reduced = pipe->count_serviced = 0;
+    pipe->my_count      = pipe->my_offset      = 0;
+
+    ucc_tl_ucp_allreduce_sliding_window_reset_buf(&pipe->accbuf);
+    for (i = 0; i < pipe->num_buffers; i++) {
+        ucc_tl_ucp_allreduce_sliding_window_reset_buf(&pipe->getbuf[i]);
+    }
+
+    for (i = 0; i < put_window_size; i++) {
+        pipe->put_requests[i] = NULL;
+    }
 }
 
 ucc_status_t
-ucc_tl_ucp_allreduce_sliding_window_start(ucc_coll_task_t __attribute__((unused)) *coll_task) //NOLINT
+ucc_tl_ucp_allreduce_sliding_window_start(ucc_coll_task_t *coll_task)
 {
-    return UCC_OK;
+    ucc_base_coll_args_t *coll_args   = &coll_task->bargs;
+    ucc_schedule_t       *schedule    = ucc_derived_of(coll_task,
+                                                       ucc_schedule_t);
+    ucc_base_team_t      *base_team   = schedule->super.team;
+    ucc_tl_ucp_team_t    *team        = ucc_derived_of(base_team,
+                                                       ucc_tl_ucp_team_t);
+    ucc_tl_ucp_context_t *tl_ctx      = UCC_TL_UCP_TEAM_CTX(team);
+    ucc_rank_t            rank        = UCC_TL_TEAM_RANK(team);
+    uint32_t              count_total = coll_task->bargs.args.dst.info.count;
+    ucc_rank_t            size        = coll_task->team->params.size;
+    ucc_datatype_t        dtype       = coll_args->args.dst.info.datatype;
+    size_t                dt_size     = ucc_dt_size(dtype);
+    int                   inplace     = UCC_IS_INPLACE(coll_args->args);
+    ucc_status_t          status      = UCC_OK;
+    int                   put_window_size = UCC_TL_UCP_TEAM_LIB(team)
+                                 ->cfg.allreduce_sliding_window_put_window_size;
+    ucc_tl_ucp_allreduce_sw_global_work_buf_info_t *gwbi_p =
+        coll_args->args.global_work_buffer;
+    ucc_tl_ucp_task_t *rdma_task = ucc_derived_of(schedule->tasks[0],
+                                                  ucc_tl_ucp_task_t);
+    ucc_tl_ucp_allreduce_sw_pipeline_t       *pipe;
+    ucc_tl_ucp_allreduce_sw_host_allgather_t *allgather_data;
+
+    pipe           = rdma_task->allreduce_sliding_window.pipe;
+    allgather_data = rdma_task->allreduce_sliding_window.allgather_data;
+
+    // Register the src buf
+    if (!inplace) {
+        status = ucc_tl_ucp_allreduce_sliding_window_register(
```

Comment:
for this i measured the difference between registering in init and registering in start. because init is called in serial by the main thread, the register would also get serialized. registering in parallel showed to be much faster

Reply:
Can you please explain how it works with multiple threads? I don't understand how you can do start in parallel and init sequential